draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1958 @@
+Unfortunately it does not meet the aforementioned requirements.
+
+`Eliot <https://github.com/itamarst/eliot>`_ is a logging library which provides
+structure and context to logs, even across coroutines, threads and processes.
+
+It is also able to emit logs to `journald`_ and has native trio integration.
+
+Network Resilience and Fault Tolerance
+--------------------------------------
+
+Celery 5 aims to be network failure resilient and fault tolerant.
+As an architectural guideline Celery must retry operations **by default**
+and must avoid doing so **indefinitely and without proper limits**.
+
+Any operation which cannot be executed either momentarily or permanently
+as a result of a bug must not be retried beyond the configured limits.
+Instead, Celery must store the operation for further inspection
+and if required, manual intervention.
+
+Celery must track and automatically handle "poisonous messages" to ensure
+the recovery of the Celery cluster.
+
+Fault Tolerance
++++++++++++++++
+
+Distributed Systems suffer from an inherent property:
+
+  Any distributed system is unreliable.
+
+  * The network may be unavailable or slow.
+  * Some or all of the servers might suffer from a hardware failure.
+  * A node in the system may arbitrarily crash
+    due to lack of memory or a bug.
+  * Any number of unaccounted failure modes.
+
+Therefore, Celery must be fault tolerant and gracefully degrade its' operation
+when failures occur.
+
+Graceful Degradation
+~~~~~~~~~~~~~~~~~~~~
+
+Features which are less mission-critical may fail at any time, provided that
+a warning is logged.
+
+This document will highlight such features and describe what happens when
+they fail for any reason.
+
+Retries
+~~~~~~~
+
+In previous Celery versions tasks were not retried by default.
+
+This forces new adopters to carefully read our documentation to ensure
+the fault tolerance of their tasks.
+
+In addition, our retry policy was declared at the task level.
+When using :ref:`celery4:task-autoretry` Celery automatically retries tasks
+when specific exceptions are raised.
+
+However the same type of exception may hold a different meaning in different
+contexts.
+
+This created the following pattern:
+
+.. code-block:: python
+
+  from celery import task
+  from data_validation_lib import validate_data
+
+  def _calculate(a, b):
+    # Do something
+
+  @task(autoretry_for=(ValueError,))
+  def complex_calculation(a, b):
+    try:
+      # Code that you don't control can raise a ValueError.
+      validate_data(a, b)
+    except ValueError:
+      print("Complete failure!")
+      return
+
+    # May temporarily raise a ValueError due to some externally fetched
+    # data which is currently incorrect but will be updated later.
+    _calculate()
+
+An obvious way around this problem is to ensure that `_calculate()`
+raises a custom exception.
+
+But we shouldn't force the users to use workarounds. Our code should be
```

Comment:
Are you sure Celery should do this? In my opinion it deviates from Celery's purpose. I personally use the following library for this purpose.  
https://github.com/jd/tenacity

Reply:
I was also thinking about tenacity

Reply:
We should encourage best practices.
Under the hood we can definitely use tenacity.